docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+This will include defining a simple protobuf type for the KV pairs. In addition to the key and value fields this message
+will include the StoreKey for the originating KVStore so that we can write out from separate KVStores to the same stream/file
+and determine the source of each KV pair.
+
+```protobuf
+message StoreKVPair {
+  optional string store_key = 1;
+  required bytes key = 2;
+  required bytes value = 3;
+}
+```
+
+```go
+// StoreKVPairWriteListener is used to configure listening to a KVStore by writing out length-prefixed
+// protobuf encoded StoreKVPairs to an underlying io.Writer
+type StoreKVPairWriteListener struct {
+	writer io.Writer
+	marshaler codec.BinaryMarshaler
+}
+
+// NewStoreKVPairWriteListener wraps creates a StoreKVPairWriteListener with a provdied io.Writer and codec.BinaryMarshaler
+func NewStoreKVPairWriteListener(w io.Writer, m codec.BinaryMarshaler) *StoreKVPairWriteListener {
+	return &StoreKVPairWriteListener{
+		writer: w,
+		marshaler: m,
+	}
+}
+
+// OnWrite satisfies the WriteListener interface by writing length-prefixed protobuf encoded StoreKVPairs
+func (wl *StoreKVPairWriteListener) OnWrite(storeKey types.StoreKey, key []byte, value []byte) {
+	kvPair := new(types.StoreKVPair)
+	kvPair.StoreKey = storeKey.Name()
+	kvPair.Key = key
+	kvPair.Value = value
+	if by, err := wl.marshaler.MarshalBinaryLengthPrefixed(kvPair); err == nil {
+		wl.writer.Write(by)
+	}
+}
+```
+
+### ListenKVStore
+
+We will create a new `Store` type `listenkv.Store` that the `MultiStore` wraps around a `KVStore` to enable state listening.
+We can configure the `Store` with a set of `WriteListener`s which stream the output to specific destinations.
+
+```go
+// Store implements the KVStore interface with listening enabled.
+// Operations are traced on each core KVStore call and written to any of the
+// underlying listeners with the proper key and operation permissions
+type Store struct {
+	parent    types.KVStore
+	listeners []types.WriteListener
+	parentStoreKey types.StoreKey
+}
+
+// NewStore returns a reference to a new traceKVStore given a parent
+// KVStore implementation and a buffered writer.
+func NewStore(parent types.KVStore, psk types.StoreKey, listeners []types.WriteListener) *Store {
+	return &Store{parent: parent, listeners: listeners, parentStoreKey: psk}
+}
+
+// Set implements the KVStore interface. It traces a write operation and
+// delegates the Set call to the parent KVStore.
+func (s *Store) Set(key []byte, value []byte) {
+	types.AssertValidKey(key)
+	s.parent.Set(key, value)
+	s.onWrite(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (s *Store) Delete(key []byte) {
+	s.parent.Delete(key)
+	s.onWrite(key, nil)
+}
+
+// onWrite writes a KVStore operation to all of the WriteListeners
+func (s *Store) onWrite(key, value []byte) {
+	for _, l := range s.listeners {
+		l.OnWrite(s.parentStoreKey, key, value)
```

Comment:
Doing this sync inside the Store means we must guarantee the listeners are fast.
Thus, disk writes may be too slow? I see the utility of not adding go-routines here, but listeners may want to just queue the data and let some other listener handle it in a goroutine.

Reply:
(answered below)

Reply:
Going to think on this a bit more. We use a goroutine to queue the data but writing it out is still currently blocking in the BaseApp abci.go method.

Reply:
Ahh actually it's not really specified if it is blocking or not right now e.g. https://github.com/cosmos/cosmos-sdk/pull/8012/files#diff-96f43642682fd41beb81b2dad576c2149ae2a020389833462e7509775c4f51cbR302, will fill in some details. We could have options for either a synchronous or asynchronous `FileStreamingService`.

Reply:
An issue I'm seeing with writing the data out asynchronously is that if that process cannot keep up with the head, then the amount of data we are holding in memory can grow indefinitely. If it can keep up with head, then I think it might as well be done synchronously.